File: src/changelog_parsing.rs
use crate::utils::IsBlank;
use anyhow::{bail, Context, Result};
use regex::Regex;
use std::borrow::Cow;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;
use std::sync::LazyLock;

pub fn get_changelog_entry(changelog_file_path: &Path, version_to_find: &str) -> Result<String> {
    let file = File::open(changelog_file_path).with_context(|| {
        format!("Failed to open changelog file at path '{changelog_file_path:?}'")
    })?;
    let reader = BufReader::new(file);

    let mut changelog_entry_lines = Vec::<String>::new();
    let mut found_changelog_entry = false;
    let target_titles = get_target_titles(version_to_find);

    let mut line_iterator = reader.lines();
    while let Some(line_result) = line_iterator.next() {
        let line = line_result.context("Failed to read line from changelog file")?;

        if !found_changelog_entry {
            if target_titles.iter().any(|title| line.starts_with(title)) {
                found_changelog_entry = true;

                match line_iterator.next() {
                    Some(Ok(next_line)) => {
                        // If line after title is blank, we don't want to include it in the
                        // changelog
                        if !next_line.is_blank() {
                            changelog_entry_lines.push(next_line)
                        }
                    }
                    Some(Err(err)) => {
                        return Err(err).context("Failed to read line from changelog file");
                    }
                    None => {
                        bail!("Unexpected end of changelog file after title line '{line}'");
                    }
                }
            }

            continue;
        }

        if changelog_entry_ended(&line) {
            break;
        }

        changelog_entry_lines.push(line);
    }

    if !found_changelog_entry {
        let [target_title_1, target_title_2] = target_titles;
        bail!("No changelog entry found for version '{version_to_find}' in changelog file '{changelog_file_path:?}' (looking for titles starting with '{target_title_1}' or '{target_title_2}')");
    }

    // Remove trailing blank lines from changelog
    while changelog_entry_lines
        .last()
        .is_some_and(|last_line| last_line.is_blank())
    {
        changelog_entry_lines.pop();
    }

    if changelog_entry_lines.is_empty() {
        bail!("Changelog entry for version '{version_to_find}' was empty")
    }

    Ok(changelog_entry_lines.join("\n"))
}

fn get_target_titles(version_to_find: &str) -> [String; 2] {
    let version_without_prefix: &str;
    let version_with_prefix: Cow<str>;

    match version_to_find.strip_prefix('v') {
        Some(version) => {
            version_without_prefix = version;
            version_with_prefix = Cow::Borrowed(version_to_find);
        }
        None => {
            version_without_prefix = version_to_find;
            version_with_prefix = Cow::Owned(format!("v{version_to_find}"));
        }
    }

    [
        format!("## [{version_without_prefix}]"),
        format!("## [{version_with_prefix}]"),
    ]
}

/// A changelog entry has ended if we find:
/// - A higher-level title (#)
/// - A new changelog entry at the same title level (##)
/// - The start of the link section at the end of the changelog
///   - Example: [v0.1.0]: <link>
fn changelog_entry_ended(line: &str) -> bool {
    static TAG_LINK_REGEX: LazyLock<Regex> = LazyLock::new(||
            // Regex:
            // - Leading ^ to match beginning of line
            // - \[ and \] to match square brackets around link text
            // - [^\[\]]+ to match link text: all characters _except_ [ or ]
            // - : to match trailing colon
            Regex::new(r#"^\[[^\[\]]+\]:"#).expect("Should be valid regex"));

    line.starts_with("# ") || line.starts_with("## ") || TAG_LINK_REGEX.is_match(&line)
}

File: internal/changelogrelease/changelog_parsing.go
package changelogrelease

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"

	"hermannm.dev/wrap"
)

func getChangelogEntry(changelogPath string, versionToFind string) (string, error) {
	absolutePath, err := filepath.Abs(changelogPath)
	if err != nil {
		return "", wrap.Errorf(
			err, "Failed to get absolute path for changelog file path '%s'", changelogPath,
		)
	}
	file, err := os.Open(absolutePath)
	if err != nil {
		return "", wrap.Errorf(err, "Failed to open changelog file at path '%s'", changelogPath)
	}
	defer file.Close()

	var entryLines []string
	foundEntry := false
	targetTitles := getTargetTitles(versionToFind)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		if !foundEntry {
			isTargetTitle := slices.ContainsFunc(
				targetTitles,
				func(targetTitle string) bool {
					return strings.HasPrefix(line, targetTitle)
				},
			)

			if isTargetTitle {
				foundEntry = true

				// Check next line - if it's blank, we don't want to include it in the changelog
				if scanner.Scan() {
					nextLine := scanner.Text()
					if nextLine != "" {
						entryLines = append(entryLines, nextLine)
					}
				}
			}

			continue
		}

		if changelogEntryEnded(line) {
			break
		}

		entryLines = append(entryLines, line)
	}
	if err := scanner.Err(); err != nil {
		return "", wrap.Error(err, "Error while reading changelog file")
	}

	if !foundEntry {
		return "", fmt.Errorf(
			"No changelog entry found for version '%s' in changelog file '%s' (looking for titles starting with one of: %v)",
			versionToFind, changelogPath, strings.Join(targetTitles, ", "),
		)
	}

	// Remove trailing blank lines from changelog
	for i, line := range slices.Backward(entryLines) {
		if line == "" {
			entryLines = slices.Delete(entryLines, i, i+1)
		}
	}

	if len(entryLines) == 0 {
		return "", fmt.Errorf("Changelog entry for version '%s' was empty", versionToFind)
	}

	return strings.Join(entryLines, "\n"), nil
}

func getTargetTitles(targetVersion string) []string {
	var versionWithPrefix string
	var versionWithoutPrefix string

	if strings.HasPrefix(targetVersion, "v") {
		versionWithPrefix = targetVersion
		versionWithoutPrefix = strings.TrimPrefix(targetVersion, "v")
	} else {
		versionWithoutPrefix = targetVersion
		versionWithPrefix = "v" + targetVersion
	}

	return []string{
		"## [" + versionWithPrefix + "]",
		"## [" + versionWithoutPrefix + "]",
	}
}

// A changelog entry has ended if we find:
// - A higher-level title (#)
// - A new changelog entry at the same title level (##)
// - The start of the link section at the end of the changelog
//   - Example: [v0.1.0]: <link>
func changelogEntryEnded(line string) bool {
	return strings.HasPrefix(line, "# ") ||
		strings.HasPrefix(line, "## ") ||
		tagLinkRegex.MatchString(line)
}

// Regex:
// - Leading ^ to match beginning of line
// - \[ and \] to match square brackets around link text
// - [^\[\]]+ to match link text: all characters _except_ [ or ]
// - : to match trailing colon
var tagLinkRegex = regexp.MustCompile(`^\[[^\[\]]+\]:`)
